File: services/agents/src/schemas/tenants.py
"""Pydantic mirror of tenants/tenant_configs tables from packages/db/src/schema/tenants.ts.

These models MUST stay in sync with the Drizzle ORM schema.
The schema contract test validates this alignment.
"""

from datetime import datetime
from enum import Enum
from typing import Any
from pydantic import BaseModel, Field


class TenantStatus(str, Enum):
    ONBOARDING = "onboarding"
    PROVISIONING = "provisioning"
    ACTIVE = "active"
    SUSPENDED = "suspended"
    DECOMMISSIONED = "decommissioned"


class Tenant(BaseModel):
    """Mirror of tenants table."""

    id: str = Field(..., description="UUID v7 primary key")
    name: str
    slug: str
    status: TenantStatus = TenantStatus.ONBOARDING
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")

    model_config = {"populate_by_name": True}


class TenantConfig(BaseModel):
    """Mirror of tenant_configs table."""

    id: str = Field(..., description="UUID v7 primary key")
    tenant_id: str = Field(..., alias="tenantId")
    config_type: str = Field(..., alias="configType")
    config_data: dict[str, Any] = Field(..., alias="configData")
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")

    model_config = {"populate_by_name": True}

File: packages/db/src/schema/tenants.ts
import { pgTable, text, timestamp, jsonb, uuid, uniqueIndex, index, check } from 'drizzle-orm/pg-core';
import { sql } from 'drizzle-orm';
import { v7 as uuid7 } from 'uuid';

export const tenantStatusEnum = ['onboarding', 'provisioning', 'active', 'suspended', 'decommissioned'] as const;
export type TenantStatus = (typeof tenantStatusEnum)[number];

export const tenants = pgTable(
  'tenants',
  {
    id: uuid('id').primaryKey().$defaultFn(() => uuid7()),
    name: text('name').notNull(),
    slug: text('slug').notNull().unique(),
    status: text('status').notNull().$type<TenantStatus>().$defaultFn(() => 'onboarding' as TenantStatus),
    createdAt: timestamp('created_at', { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp('updated_at', { withTimezone: true }).notNull().defaultNow().$onUpdateFn(() => new Date()),
  },
  (table) => [
    check('chk_tenant_status', sql`${table.status} IN ('onboarding', 'provisioning', 'active', 'suspended', 'decommissioned')`),
  ],
);

export const tenantConfigs = pgTable(
  'tenant_configs',
  {
    id: uuid('id').primaryKey().$defaultFn(() => uuid7()),
    tenantId: uuid('tenant_id')
      .notNull()
      .references(() => tenants.id, { onDelete: 'cascade' }),
    configType: text('config_type').notNull(),
    configData: jsonb('config_data').notNull(),
    createdAt: timestamp('created_at', { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp('updated_at', { withTimezone: true }).notNull().defaultNow().$onUpdateFn(() => new Date()),
  },
  (table) => [
    uniqueIndex('uq_tenant_configs_tenant_type').on(table.tenantId, table.configType),
    index('idx_tenant_configs_config_data').using('gin', table.configData),
  ],
);
